components/style/matching.rs

```diff
@@ -821,81 +828,7 @@ pub trait MatchMethods : TElement {
-            // so that we can start the binding load.  Otherwise, there is no
-            // need for damage.
-            return RestyleDamage::compute_undisplayed_style_difference(old_values, new_values);
-        }
-
-        if pseudo.map_or(false, |p| p.is_before_or_after()) {
-            // FIXME(bz) This duplicates some of the logic in
-            // PseudoElement::should_exist, but it's not clear how best to share
-            // that logic without redoing the "get the display" work.
-            let old_style_generates_no_pseudo =
-                old_style_is_display_none ||
-                old_values.ineffective_content_property();
-
-            let new_style_generates_no_pseudo =
-                new_style_is_display_none ||
-                new_values.ineffective_content_property();
-
-            if old_style_generates_no_pseudo != new_style_generates_no_pseudo {
-                return StyleDifference::new(RestyleDamage::reconstruct(), StyleChange::Changed)
-            }
-
-            // The pseudo-element will remain undisplayed, so just avoid
-            // triggering any change.
-            //
-            // NOTE(emilio): We will only arrive here for pseudo-elements that
-            // aren't generated (see the is_existing_before_or_after check in
-            // accumulate_damage).
-            //
-            // However, it may be the case that the style of this element would
-            // make us think we need a pseudo, but we don't, like for pseudos in
-            // replaced elements, that's why we need the old != new instead of
-            // just check whether the new style would generate a pseudo.
-            return StyleDifference::new(RestyleDamage::empty(), StyleChange::Unchanged)
-        }
-
-        if pseudo.map_or(false, |p| p.is_first_letter() || p.is_first_line()) {
-            // No one cares about this pseudo, and we've checked above that
-            // we're not switching from a "cares" to a "doesn't care" state
-            // or vice versa.
-            return StyleDifference::new(RestyleDamage::empty(),
-                                        StyleChange::Unchanged)
-        }
-
-        // If we are changing display property we need to accumulate
-        // reconstruction damage for the change.
-        // FIXME: Bug 1378972: This is a workaround for bug 1374175, we should
-        // generate more accurate restyle damage in fallback cases.
-        let needs_reconstruction = new_display != old_display;
-        let damage = if needs_reconstruction {
-            RestyleDamage::reconstruct()
-        } else {
-            RestyleDamage::empty()
-        };
-        // We don't really know if there was a change in any style (since we
-        // didn't actually call compute_style_difference) but we return
-        // StyleChange::Changed conservatively.
-        StyleDifference::new(damage, StyleChange::Changed)
+        RestyleDamage::compute_style_difference(old_values, new_values)
```

Comment:
We... don't, because we don't need to.

The no existing style case was kind of a hack because we weren't able to find the old style context in the frame tree, which was needed to get the right bits. Now that's gone, we can also remove the `display_changed` hack, which fixes bug 1378972.